File: scripts/fetch-gsc-data.py
#!/usr/bin/env python3
"""
Script para extrair dados do Google Search Console para o RADAR v2

Uso:
  python fetch-gsc-data.py --site "sc-domain:bb.com.br" --days 90 --output radar-data.csv

Requer:
  - Google Cloud Project com Search Console API enabled
  - Service Account JSON credentials em ~/.config/gsc-credentials.json
  - Site verificado no GSC com Service Account added
"""

import argparse
import csv
import json
from datetime import datetime, timedelta
from pathlib import Path

try:
    from google.oauth2 import service_account
    from googleapiclient.discovery import build
except ImportError:
    print("❌ Instale: pip install google-auth google-api-python-client")
    exit(1)


CREDENTIALS_PATH = Path.home() / ".config" / "gsc-credentials.json"


def get_credentials():
    """Load Service Account credentials"""
    if not CREDENTIALS_PATH.exists():
        raise FileNotFoundError(
            f"Credentials not found at {CREDENTIALS_PATH}\n"
            "Create a Service Account in Google Cloud Console and save JSON here"
        )
    return service_account.Credentials.from_service_account_file(
        str(CREDENTIALS_PATH),
        scopes=["https://www.googleapis.com/auth/webmasters.readonly"]
    )


def fetch_gsc_data(site_url: str, start_date: str, end_date: str, row_limit: int = 5000):
    """Fetch search analytics data from GSC API"""
    creds = get_credentials()
    service = build("searchconsole", "v1", credentials=creds)

    request = {
        "startDate": start_date,
        "endDate": end_date,
        "dimensions": ["query"],
        "aggregationType": "auto",
        "rowLimit": row_limit,
        "type": "web"
    }

    response = service.searchanalytics().query(siteUrl=site_url, body=request).execute()

    if "rows" not in response:
        print(f"⚠️ No data returned. Check site verification and permissions.")
        return []

    return response["rows"]


def transform_for_radar(rows: list) -> list:
    """Transform GSC API response to RADAR format"""
    data = []
    for row in rows:
        term = row["keys"][0]
        clicks = row["clicks"]
        impressions = row["impressions"]
        ctr = row["ctr"] * 100  # Convert decimal to percentage
        position = row["position"]
        
        data.append({
            "term": term,
            "clicks": clicks,
            "impressions": impressions,
            "ctr": round(ctr, 2),
            "position": round(position, 2)
        })
    
    return data


def write_csv(data: list, output_path: str):
    """Write data to CSV file"""
    fieldnames = ["term", "clicks", "impressions", "ctr", "position"]
    
    with open(output_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(data)
    
    print(f"✅ Wrote {len(data)} rows to {output_path}")


def main():
    parser = argparse.ArgumentParser(description="Fetch GSC data for RADAR v2")
    parser.add_argument("--site", required=True, help="Site URL (e.g., 'sc-domain:bb.com.br')")
    parser.add_argument("--days", type=int, default=90, help="Number of days to fetch")
    parser.add_argument("--output", default="radar-data.csv", help="Output CSV file")
    parser.add_argument("--limit", type=int, default=5000, help="Max rows to fetch")
    
    args = parser.parse_args()
    
    # Calculate date range
    end_date = datetime.now().strftime("%Y-%m-%d")
    start_date = (datetime.now() - timedelta(days=args.days)).strftime("%Y-%m-%d")
    
    print(f"🔍 Fetching GSC data for {args.site}")
    print(f"   Date range: {start_date} to {end_date}")
    print(f"   Row limit: {args.limit}")
    
    # Fetch data
    rows = fetch_gsc_data(args.site, start_date, end_date, args.limit)
    
    if not rows:
        return
    
    # Transform
    data = transform_for_radar(rows)
    
    # Write CSV
    write_csv(data, args.output)
    
    # Print summary
    total_clicks = sum(row["clicks"] for row in data)
    total_impressions = sum(row["impressions"] for row in data)
    avg_ctr = (total_clicks / total_impressions * 100) if total_impressions > 0 else 0
    
    print(f"\n📊 Summary:")
    print(f"   Total terms: {len(data)}")
    print(f"   Total clicks: {total_clicks:,}")
    print(f"   Total impressions: {total_impressions:,}")
    print(f"   Average CTR: {avg_ctr:.2f}%")


if __name__ == "__main__":
    main()

File: scripts/fetch-gsc-data.js
#!/usr/bin/env node
/**
 * Script para extrair dados do Google Search Console para o RADAR v2
 * 
 * Uso:
 *   node fetch-gsc-data.js --site "sc-domain:bb.com.br" --days 90 --output radar-data.csv
 * 
 * Requer:
 *   - npm install googleapis
 *   - Service Account JSON em ~/.config/gsc-credentials.json
 */

const { google } = require('googleapis');
const fs = require('fs');
const path = require('path');
const { parseArgs } = require('util');

const CREDENTIALS_PATH = path.join(process.env.HOME, '.config', 'gsc-credentials.json');

// Parse command line arguments
const { values } = parseArgs({
  options: {
    site: { type: 'string', short: 's' },
    days: { type: 'string', short: 'd', default: '90' },
    output: { type: 'string', short: 'o', default: 'radar-data.csv' },
    limit: { type: 'string', short: 'l', default: '5000' },
  },
  strict: false
});

if (!values.site) {
  console.error('❌ --site required. Example: --site "sc-domain:bb.com.br"');
  process.exit(1);
}

async function fetchGSCData() {
  // Load credentials
  if (!fs.existsSync(CREDENTIALS_PATH)) {
    console.error(`❌ Credentials not found at ${CREDENTIALS_PATH}`);
    console.error('Create a Service Account in Google Cloud Console and save JSON there');
    process.exit(1);
  }

  const auth = new google.auth.GoogleAuth({
    keyFile: CREDENTIALS_PATH,
    scopes: ['https://www.googleapis.com/auth/webmasters.readonly']
  });

  const searchconsole = google.searchconsole({ version: 'v1', auth });

  // Calculate date range
  const days = parseInt(values.days) || 90;
  const endDate = new Date().toISOString().split('T')[0];
  const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString().split('T')[0];
  const rowLimit = parseInt(values.limit) || 5000;

  console.log(`🔍 Fetching GSC data for ${values.site}`);
  console.log(`   Date range: ${startDate} to ${endDate}`);
  console.log(`   Row limit: ${rowLimit}`);

  try {
    const response = await searchconsole.searchanalytics.query({
      siteUrl: values.site,
      requestBody: {
        startDate,
        endDate,
        dimensions: ['query'],
        aggregationType: 'auto',
        rowLimit,
        type: 'web'
      }
    });

    const rows = response.data.rows || [];

    if (rows.length === 0) {
      console.log('⚠️ No data returned. Check site verification and permissions.');
      return;
    }

    // Transform to RADAR format
    const data = rows.map(row => ({
      term: row.keys[0],
      clicks: row.clicks,
      impressions: row.impressions,
      ctr: Math.round(row.ctr * 100, 2),  // GSC returns decimal, convert to %
      position: Math.round(row.position, 2)
    }));

    // Write CSV
    const csv = [
      'term,clicks,impressions,ctr,position',
      ...data.map(row => `${row.term},${row.clicks},${row.impressions},${row.ctr},${row.position}`)
    ].join('\n');

    fs.writeFileSync(values.output, csv);
    console.log(`✅ Wrote ${data.length} rows to ${values.output}`);

    // Summary
    const totalClicks = data.reduce((sum, row) => sum + row.clicks, 0);
    const totalImpressions = data.reduce((sum, row) => sum + row.impressions, 0);
    const avgCTR = totalImpressions > 0 ? (totalClicks / totalImpressions * 100) : 0;

    console.log('\n📊 Summary:');
    console.log(`   Total terms: ${data.length}`);
    console.log(`   Total clicks: ${totalClicks.toLocaleString()}`);
    console.log(`   Total impressions: ${totalImpressions.toLocaleString()}`);
    console.log(`   Average CTR: ${avgCTR.toFixed(2)}%`);

  } catch (error) {
    console.error('❌ Error fetching data:', error.message);
    process.exit(1);
  }
}

fetchGSCData();